tracer/src/Datadog.Trace/Agent/MessagePack/SpanMessagePackFormatter.cs

```diff
@@ -201,6 +214,82 @@ private int Serialize(ref byte[] bytes, int offset, in SpanModel spanModel)
                 offset += WriteMetaStruct(ref bytes, offset, in spanModel);
             }
 
+            if (span.SpanLinks is { Count: > 0 })
+            {
+                offset += WriteSpanLink(ref bytes, offset, in spanModel);
+            }
+
+            return offset - originalOffset;
+        }
+
+        private int WriteSpanLink(ref byte[] bytes, int offset, in SpanModel spanModel)
+        {
+            int originalOffset = offset;
+            offset += MessagePackBinary.WriteStringBytes(ref bytes, offset, _spanLinkBytes);
+            offset += MessagePackBinary.WriteArrayHeader(ref bytes, offset, spanModel.Span.SpanLinks.Count);
+            foreach (var spanLink in spanModel.Span.SpanLinks)
+            {
+                var context = spanLink.Context;
+                var traceState = W3CTraceContextPropagator.CreateTraceStateHeader(context);
```

Comment:
Reading the RFC I'm a bit confused as to why `tracestate` is listed as an optional attribute, so I'm thinking that it is intended to only be added when this `SpanLink` (or maybe `Span`) came from a distributed context?

For what it is worth, `W3CTraceContextPropagator.CreateTraceStateHeader(context);` won't ever return `null` or an `empty` string even when the link/span aren't from a distributed context. 

In one of my [PR](https://github.com/DataDog/dd-trace-dotnet/pull/5176/files#diff-8bc1e22eec771ec585cd5288fea93aca9c32ba0bdaba321683d3d2e09c5fdd95)'s I do have an `IsRemote` added to `SpanContext` which may be useful here to wrap the call `CreateTraceStateHeader(context)` with.

@lucaspimentel thoughts?

Reply:
I'm honestly unsure myself - I'll wait for inputs from others.

Reply:
<img width="283" alt="image" src="https://github.com/DataDog/dd-trace-dotnet/assets/73513662/a6163fbf-c6a8-478d-85c6-6e5e48c8d86d">

So checking the IsRemote before adding the TraceState seems to be the call

Reply:
I was originally planning to wrap 
`                var traceState = W3CTraceContextPropagator.traceState(context);`
under `(isRemote)` check but the scoping of `traceState` in the block of serializing/increase len gets messed up. 

I just moved it so `W3CTraceContextPropagator` is always called, but we the decision to serialize and allocate space for the serialization is under `isRemote`

removed isNullOrEmpty since it's never null or empty.

Reply:
@bouwkast did you mean #5385 for adding `IsRemote`?

Reply:
Yes thanks, I actually extracted it out (and modified slightly) of that first PR so it could be rebased on this one